### ISSUE TITLE
Incorrect handling of output parameter when using subparser

### ISSUE THREAD

**Issue by jderuiter** [p4lang/p4c#1879]
To simplify some code that I'm writing I started to use a subparser. The subparser currently consists of only one state and looks roughly as follows:
```
parser SubParser(packet_in packet, out subheader_t sub, inout metadata meta, in bit<8> selectedPos) {
    state start {
        packet.extract(sub);

        bool currentFieldSelected = selectedPos == meta.currPos;
        meta.currentVal.value = meta.currentVal.value + (bit<1>)currentFieldSelected * sub.value;
        meta.currPos = meta.currPos + 1;
        transition accept;
    }
}
```
When compiling this code seems to be simplified in an incorrect way however. When I run this code with `simple_switch` and `currentFieldSelected` evaluates to true, I see the following line in the debug output:
```
Parser set: setting field 'meta_0_currentVal.value from expression, new value is 1
```
This is correct, but then a bit later (presumably at the end of the subparser) there is the following line:
```
Parser set: setting field 'scalars.metadata._currentVal_value7' from expression, new value is 0
```
Instead of copying the previously computed value, the expression seems to be evaluated again, resulting in an incorrect value (as meta.currPos was updated in the meantime).

If i move the condition outside the parser and pass it as an argument, it does function correctly and I get the following line in the output instead of the one above:
```
Parser set: setting field 'scalars.metadata._currentVal_value7' from field 'meta_0_currentVal.value' (1)
```
So now the computed value is actually assigned and the expression not evaluated again.

**Comment by jafingerhut** [comment]
Do you have a complete P4 program that exhibits the incorrect behavior you have seen, that can be attached to this issue (i.e. shared publicly with the world)?

**Comment by jderuiter** [comment]
I tried to construct a minimal example from scratch that exhibits this behaviour, but did not succeed yet. The code I have this issue with is slightly complex so probably not very nice for debugging. I'll try to strip down the code to see whether I can get something somewhat readable.

**Comment by jafingerhut** [comment]
Thanks for the attempt at reducing the example size -- it is appreciated.  Please do not "lose" the buggy version during your efforts, and if it is sharable, someone else may be able to reduce its size while still exhibiting the bug, too.

**Comment by jderuiter** [comment]
Attached you find a stripped down version of my code and a pcap containing the packet I used for testing. The problem is with `meta.currenti.upDirection` in the subparser.

[prot.tar.gz](https://github.com/p4lang/p4c/files/3108357/prot.tar.gz)

**Comment by mbudiu-vmw** [comment]
Can you try to convert the pcap file into an STF file? The format is documented here: https://github.com/p4lang/p4c/blob/master/docs/compiler-design.pptx
Ideally we would like the packet to pass only if the compilation is successful.

You can see what the compiler has done to your program after each pass. For example, to see what it looks like after the front-end I used `./p4c-bm2-ss --top4 FrontEndLast prot.p4` I couldn't notice an incorrect translation, but the program is fairly complicated, so I may have missed something. The list of all passes can be printed by adding -v to the compilation command.

**Comment by jderuiter** [comment]
The final JSON at least shows a difference between the two version (left correct, right incorrect):
```
1342c1342
<                                             "value" : ["scalars", "metadata._currPos3"]
---
>                                             "value" : ["prot_common", "curri"]
1346c1346
<                                             "value" : ["prot_common", "curri"]
---
>                                             "value" : ["scalars", "metadata._currPos3"]
1486,1487c1486,1554
<                   "type" : "field",
<                   "value" : ["meta_0_currenti", "upDirection"]
---
>                   "type" : "expression",
>                   "value" : {
>                     "type" : "expression",
>                     "value" : {
>                       "op" : "&",
>                       "left" : {
>                         "type" : "expression",
>                         "value" : {
>                           "op" : "+",
>                           "left" : {
>                             "type" : "field",
>                             "value" : ["scalars", "metadata._currenti_upDirection4"]
>                           },
>                           "right" : {
>                             "type" : "expression",
>                             "value" : {
>                               "op" : "&",
>                               "left" : {
>                                 "type" : "expression",
>                                 "value" : {
>                                   "op" : "*",
>                                   "left" : {
>                                     "type" : "expression",
>                                     "value" : {
>                                       "op" : "?",
>                                       "left" : {
>                                         "type" : "hexstr",
>                                         "value" : "0x01"
>                                       },
>                                       "right" : {
>                                         "type" : "hexstr",
>                                         "value" : "0x00"
>                                       },
>                                       "cond" : {
>                                         "type" : "expression",
>                                         "value" : {
>                                           "op" : "==",
>                                           "left" : {
>                                             "type" : "field",
>                                             "value" : ["prot_common", "curri"]
>                                           },
>                                           "right" : {
>                                             "type" : "field",
>                                             "value" : ["scalars", "metadata._currPos3"]
>                                           }
>                                         }
>                                       }
>                                     }
>                                   },
>                                   "right" : {
>                                     "type" : "field",
>                                     "value" : ["inf_0", "upDirection"]
>                                   }
>                                 }
>                               },
>                               "right" : {
>                                 "type" : "hexstr",
>                                 "value" : "0x01"
>                               }
>                             }
>                           }
>                         }
>                       },
>                       "right" : {
>                         "type" : "hexstr",
>                         "value" : "0x01"
>                       }
>                     }
>                   }
```

**Comment by jderuiter** [comment]
Attached you find the packet in STF format and a version of the code that will drop the incorrect packet.

[prot.p4.txt](https://github.com/p4lang/p4c/files/3111062/prot.p4.txt)
[prot.stf.txt](https://github.com/p4lang/p4c/files/3111063/prot.stf.txt)

**Comment by jderuiter** [comment]
I narrowed down the issue to the `BMV2::SimpleSwitchMidEnd_27_LocalCopyPropagation` step. Attached you find the intermediate output for this step. The offending code is on line 1883. Instead of copying the value computed on line 1813, the expression is evaluated again but in `meta._currPos3` has been updated so the result is not the same as before.

[scion-BMV2::SimpleSwitchMidEnd_27_LocalCopyPropagation.p4.txt](https://github.com/p4lang/p4c/files/3111246/scion-BMV2.SimpleSwitchMidEnd_27_LocalCopyPropagation.p4.txt)

**Comment by jderuiter** [comment]
The problem seems to be in `dropValuesUsing(meta._currPos3)`, where `meta_0_currenti.upDirection` is not dropped even though it contains `meta._currPos3`. This seems to be caused by incorrect behaviour of `exprUses`. When adding the following test for `exprUses`, the second test actually fails:
```
    auto add2 = new IR::Add(obj1_f1, obj1_f11);

    EXPECT_TRUE(exprUses(add2, "obj1.f1"));
    EXPECT_TRUE(exprUses(add2, "obj1.f11"));
```

**Comment by jderuiter** [comment]
The `exprUses.preorder` method was only called once for Path, instead of twice as you would expect (once for every side of the addition). As `search_tail` was updated in this method, the second occurrence of `obj1` was not stripped from the path and therefore the search string remained "obj1.f11" instead of just "f11" (and therefor never matched). Moving the updating of `search_tail` to a `preorder` and a `revisit` method for PathExpression seems to fix this issue (Path was not revisited).

I'll try to make a pull request for this. However, as I only started looking into the p4c code today I'm not sure this wouldn't break other things. Any comments on this would be welcome. One issue I see could be if Path is ever used outside of a PathExpression, as then `search_tail` would not be updated, but I'm not sure this ever happens.
